src/generator.rs

```diff
@@ -120,16 +117,21 @@ impl Generator {
     /// ```
     pub fn generate_from_datetime_with_source<R>(
         &mut self,
-        datetime: OffsetDateTime,
+        datetime: SystemTime,
         source: &mut R,
     ) -> Result<Ulid, MonotonicError>
     where
         R: rand::Rng,
     {
-        let last_ms = self.previous.timestamp_ms() as i128;
+        let last_ms = self.previous.timestamp_ms();
         // maybe time went backward, or it is the same ms.
         // increment instead of generating a new random so that it is monotonic
-        if (datetime.unix_timestamp_nanos() / 1_000_000) <= last_ms {
+        if datetime
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
```

Comment:
Let's not panic. Instead we can saturate at zero with: `.unwrap_or(Duration::ZERO)`.

Reply:
Good idea - I've also documented the behavior for `SystemTime`'s earlier than the unix epoch and added some tests to verify the behavior